versioning/BRM/extentmap.cpp

```diff
@@ -3253,101 +3006,85 @@ void ExtentMap::logAndSetEMIndexReadOnly(const std::string& funcName)
-
-    if (isUnsigned(colDataType))
-    {
-        e->partition.cprange.lo_val = numeric_limits<uint64_t>::max();
-        e->partition.cprange.hi_val = 0;
-    }
-    else
-    {
-        e->partition.cprange.lo_val = numeric_limits<int64_t>::max();
-        e->partition.cprange.hi_val = numeric_limits<int64_t>::min();
-    }
-
-    e->partition.cprange.sequenceNum = 0;
-
-    e->colWid       = colWidth;
-
-    e->dbRoot       = dbRoot;
-    e->partitionNum = partitionNum;
-    e->segmentNum   = segmentNum;
-    e->status       = EXTENTUNAVAILABLE; // mark extent as in process
-
-    // If first extent for this OID, partition, dbroot, and segment then
-    //   blockOffset is set to 0
-    // else
-    //   blockOffset is extrapolated from the last extent
-    if (lastExtentIndex == -1)
-    {
-        e->blockOffset  = 0;
-        e->HWM          = 0;
-    }
-    else
-    {
-        e->blockOffset  = static_cast<uint64_t>
-                          (fExtentMap[lastExtentIndex].range.size) * 1024 +
-                          fExtentMap[lastExtentIndex].blockOffset;
-        e->HWM          = 0;
-    }
-
-    // Partition, segment, and blockOffset 0 represents new table or column.
-    // When DDL creates a table, we can mark the first extent as VALID, since
-    // the table has no data.  Marking as VALID enables cpimport to update
-    // the CP min/max for the first import.
-    // If DDL is adding a column to an existing table, setting to VALID won't
-    // hurt, because DDL resets to INVALID after the extent is created.
-    if ((e->partitionNum == 0) &&
-            (e->segmentNum   == 0) &&
-            (e->blockOffset  == 0))
-        e->partition.cprange.isValid = CP_VALID;
-    else
-        e->partition.cprange.isValid = CP_INVALID;
-
-    startBlockOffset = e->blockOffset;
-
-    makeUndoRecord(fEMShminfo, sizeof(MSTEntry));
-    fEMShminfo->currentSize += sizeof(struct EMEntry);
-    auto resShmemHasGrownPair = fPExtMapIndexImpl_->insert(fExtentMap[emptyEMEntry], emptyEMEntry);
-
-    if (resShmemHasGrownPair.second)
-        fEMIndexShminfo->allocdSize = fPExtMapIndexImpl_->getShmemSize();
-
-    if (!resShmemHasGrownPair.first)
-        logAndSetEMIndexReadOnly("_createColumnExtentExactFile");
-
-    return startLBID;
+LBID_t ExtentMap::_createColumnExtentExactFile(uint32_t size, int OID, uint32_t colWidth, uint16_t dbRoot,
+                                               uint32_t partitionNum, uint16_t segmentNum,
+                                               execplan::CalpontSystemCatalog::ColDataType colDataType,
+                                               uint32_t& startBlockOffset)
+{
+  uint32_t highestOffset = 0;
+  LBID_t startLBID = getLBIDsFromFreeList(size);
+  EMEntry* lastEmEntry = nullptr;
+
+  const auto lbids = fPExtMapIndexImpl_->find(dbRoot, OID);
+  auto emIdents = getEmIdentsByLbids(lbids);
+  for (auto& emEntry : emIdents)
+  {
+    if ((emEntry.fileID == OID) && (emEntry.segmentNum == segmentNum) &&
```

Comment:
OID and segment num comparisons looks redundant. IMHO You either need to remove them or add an error processing block that throws if OID and segmentNum doesn't match.